docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,74 @@
+
+The objective with the Connection Diagnostics is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
+
+To avoid further complicating the already complex logic of the controller, this document proposes encapsulating this new functionality within the Connection Diagnostics. This object will interact with both the controller and connection health, serving as a dedicated entity for identifying connection problems and correctly communicating status changes to the UI, ultimately enhancing the user experience.
+
+## Proposed Solution
+
+The Connection Diagnostics will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionDiagnostics::State`. The Connection Diagnostics will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful. 
+
+In the event of a probe failure, the Connection Diagnostics will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted and the Connection Diagnostics will enter the Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Diagnostics.
+
+> For specific details about how these failures are handled and communicated with the user, please refer to this [Figma document](https://www.figma.com/file/GZKigVZbnCisMJNU3Hc9xM/Connection-error-messaging?type=design&node-id=26-215&mode=design&t=kfYnUKuXQ9nsGDRR-0).
+
+## Implementation Plan
+
+```c++
+enum ConnectionDiagnosticsStates {
+  StateProbingInternet,
+  StateCheckingFirewall,
+  StateProbingServerLocation,
+  StateCheckingSubscription,
+  StateIdle,
+}
+```
+
+### Upon VPN Activation (`StateOff` to `StateConnecting`) - Phase 1
+Perform a single cycle going through all the Connection Diagnostics probes.
```

Comment:
Why do we need to run these checks upon activation? Why not attempt to activate and only run the chceks if activation is unsuccesful?

Reply:
Also, for my understanding who owns ConnectionDiagnostics? Controller?

Reply:
> Why do we need to run these checks upon activation? Why not attempt to activate and only run the cheks if activation is unsuccessful?

Hmmm I'd be open to that. @lesleyjanenorton what are your thoughts?

> Also, for my understanding who owns ConnectionDiagnostics? Controller?

Not sure I understand this question, can you elaborate?

Reply:
Sure, I mean who is responsible for instantiating and managing the ConnectionDiagnostics object?

Reply:
> Sure, I mean who is responsible for instantiating and managing the ConnectionDiagnostics object?

The diagnostics tool runs when stability is `NoSignal` so I think the connection health class is responsible for instantiating the diagnostics and managing it.